File: bsm.py
"""
Black-Scholes Equations
s: spot price
k: strike price
t: expiry time in years
r: annualized risk-free interest rate
sig: volatility
c: call option price
"""
import numpy as np
from scipy.stats import norm


def call_value(s, k, t, r, sigma):
    d1 = (np.log(s / k) + (r + .5 * sigma ** 2) * t) / (sigma * t ** .5)
    d2 = d1 - sigma * t ** .5
    c = s * norm.cdf(d1) - k * np.exp(-r * t) * norm.cdf(d2)
    return c


def call_vega(s, k, t, r, sigma):
    d1 = (np.log(s / k) + (r + .5 * sigma ** 2) * t) / (sigma * t ** .5)
    vega = np.exp(-r*t) * s * t ** .5 * norm.pdf(d1)
    return vega

File: src/bsm.rs
use statrs::distribution::{Normal, ContinuousCDF};

fn norm_cdf(x: f64) -> f64 {
    let norm: Normal = Normal::new(0.0, 1.0).unwrap();
    norm.cdf(x)
}

pub fn call_value(s: f64, k:f64, t:f64, r:f64, sigma:f64) -> f64 {
    let d1 = ((s / k).ln() + (r + 0.5 * sigma.powf(2.0)) * t) / (sigma * t.sqrt());
    let d2 = d1 - sigma * t.sqrt();
    let c = s * norm_cdf(d1) - k * (- r * t).exp() * norm_cdf(d2);
    c
}

pub fn call_vega(s: f64, k:f64, t:f64, r:f64, sigma:f64) -> f64 {
    let d1 = ((s / k).ln() + (r + 0.5 * sigma.powf(2.0)) * t) / (sigma * t.sqrt());
    let vega = (- r * t).exp() * s * t.sqrt() * norm_cdf(d1);
    vega
}
